<a href="https://colab.research.google.com/github/GPrathap/pythonII/blob/master/pythonII/sec_numpy/numpy_exercises_set2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Exercise 2 - Image processing with NumPy

Various image processing can be performed using NumPy functions by reading the image as a NumPy array.

රූපය (image), NumPy array ලෙස ගැනීමෙන් විවිධ රූප සැකසුම් සිදුකිරීමට NumPy library භාවිත කල හැකිය.

## Following information will be useful to understand the applciation of NumPy in image processing

Images are compiled using pixels, or tiny dots, containing unique colour and tonal information that come together to create the image.

පින්තුර සැදී ඇත්තේ පික්සල (pixel) හෙවත් කුඩා තිත් වලිනි. මෙම pixel වල රූපය නිර්මාණය කිරීම අවශ්‍ය වන වර්ණ සහ tone (එලිය, අඳුර, වර්ණ විචලනය) පිලිබඳ තොරතුරු අඩංගු වේ.

Look at the following example.
පහත උදාහරණය බලන්න.

This is a 6x4 pixel RGB image.

පහත දැක්වෙන්නේ 6 x 4 pixel පින්තුරයකි.

<img src='https://drive.google.com/uc?id=1a4THRiwHXIPsgaGMRzDp-xiClwCeA5F4' width="300" height="200">

*   This image contains 4 lines of pixels. 
    
    මෙම රූපය pixel පේලි 4කින් සමන්විතය.

<img src='https://drive.google.com/uc?id=1aWt9_KnRpKHnntpFYoxJQmv9TcR66zzo' width="300" height="200">

*   Each line of pixels contains 6 pixels.

    සෑම pixel පේලියක්ම pixels 6කින් සමන්විතය. 

<img src='https://drive.google.com/uc?id=1kc7shJwjLpBHkza4_JblWBcKO2cGwAFH' width="300" height="80">

*   Each pixel contains 3 bytes (representing the red, green and blue values of the pixel colour).

    සෑම pixel එකකම bytes 3ක් ඇත. ඒවායින් pixel එකෙහි වර්ණයේ රතු, කොළ, සහ නිල් වර්ණ සංයුතීන් නිරුපණය වේ.

<img src='https://drive.google.com/uc?id=17IVXNvdd3kapiNc-ciEtPsqOIEG29leq' width="120" height="60">

Therefore, RGB images are usually stored as 3-dimensional arrays and the shape of the array **height x width x 3**.

එමනිසා RGB පින්තුර මාන 3ක arrays වල ගබඩා කරනු ලැබේ. එම arrays වල හැඩය **උස x පළල x 3** වේ.

Here is how we create an array to represent a 6x4 pixel RGB image.

පහතින් දැක්වෙන්නේ 6x4 pixel RGB පින්තුරයක් නිරුපණය කිරීමට array එකක් නිර්මාණය කරන ආකාරයයි.








In [ ]:
import numpy as np

width = 5
height = 4

array = np.zeros([height, width, 3], dtype=np.uint8)


 
**Note:**

The data is ordered by lines, pixels for each line, and then 3-byte values for RGB (pixel details). Therefore, the first dimension is the height, and the second dimension is the width. 

array එක නිර්මාණයේදී මුලින් පින්තුරයේ ඇති pixel පේලි ගණනද දෙවනුව එක පේළියක ඇති pixels ගණනද අවසානයේ RGB සංයුතීන් නිරුපණයට byte 3ක් ද ලබා දිය යුතුය.


##Import required libraries

In [3]:
from PIL import Image #Python Imaging Library
import numpy as np

## Download the dataset

Let's download the necessary datasets to your working environment.

අවශ්‍ය දත්ත කට්ටල බාගත කරමු.

In [11]:
from urllib.request import urlopen   

URL = 'https://raw.githubusercontent.com/GPrathap/pythonII/master/dataset/cats/cat_00/00000002_001.jpg?raw=true'
url = urlopen(URL)
output_img = open('/content/sample_data/cat.jpg', 'wb')    # note the flag:  "wb"        
output_img.write(url.read())
output_img.close()

##Let's start the exercise

###Read the image file as NumPy array

Original image:

<img src='https://drive.google.com/uc?id=18k1JMNFypRODuEgeqpMX_M1jhh0_zP2R' width="300" height="250">

In [17]:
image_file='/content/sample_data/cat.jpg'
img_array = np.array(Image.open(image_file))

Let's look at the data type and the shape of the array.

We used **dtype** and **shape** attributes to find the data type and the shape of the array respectively.


In [18]:
print(img_array.dtype)
print(img_array.shape)

uint8
(768, 1024, 3)


According to the above results, this array is a **3-dimentional** array with **uint8** data type. The array has 768 rows and 1024 columns. This implies that the image has 768 pixel lines and 1024 pixels for each line.

**Note:**
uint8 is an 8-bit unsigned integer (range: 0 through 255 decimal).

###Get and set (change) pixel values

You can get the value of a pixel by specifying the coordinates at the index [row, column] of array.

In [20]:
print(img_array[100, 200])

[68 58 57]


The 200th pixel in the 100th line has intensity values of 68, 58 and 57 for red, green, and blue respectively.

You can also assign RGB values of a specific pixel to separate variables.

In [21]:
red, green, blue = img_array[100, 200]

print("Red intensity: ",red)
print("Green intensity: ",green)
print("Blue intensity: ",blue)

Red intensity:  68
Green intensity:  58
Blue intensity:  57


It is also possible to get the value by specifying the color.

In [22]:
print("Red intensity: ",img_array[100, 200, 0])
print("Green intensity: ",img_array[100, 200, 1])
print("Blue intensity: ",img_array[100, 200, 2])

Red intensity:  68
Green intensity:  58
Blue intensity:  57


You can also change pixel value to a new value. 

You can change RGB all at once or change a single color you want to change.

In [23]:
#change the entire pixel value
img_array[100, 200] = (0, 50, 100)
print(img_array[100, 200])


[  0  50 100]


Now the value of the pixel is not [68 58 57]. It is [  0  50 100].

In [25]:
#change only the red intensity
img_array[100, 200,0] = 120
print(img_array[100, 200])


[120  50 100]


Since we changed the red intensity, the new value of the pixel is [120  50 100]

###Let's change the image to single color image

You can generate single-color images by setting other color values to 0, and concatenate them horizontally with **np.concatenate()** to create a single image.

eg.: To convert the image into red color image, we set green and blue intensity values of each pixel to 0.

**Note:**

*   **np.concatenate()** a function from the NumPy package which combines NumPy arrays together.

<img src='https://drive.google.com/uc?id=1w1aCI4R15pdhpkyBuuWc6TO58xgKCjBQ' width="300" height="200">

*   We used **copy()** function to make a copy of a orginal array since we should not change the original numpy array.

*   Let's look at the **img_R[:, :, (1, 2)] = 0**

  Since we change the image to red color image, we want to set the green and blue intensity values of **each pixel** to 0. In Numpy, you can use **colon (:)** which means **fetch everything**. 
  
  img_R[:, :] means every pixel in the image or every rows and columns in the numpy array. 
  
  Then, we have to add the index of the color we want to change. If we use **img_R[:, :, 1] = 0** , this only changes the green intensity of all the pixels to 0. We will need another **img_R[:, :, 2] = 0** to change the blue intesities. When you use **(1, 2)**, you can change both green and blue intensity values at once.



In [31]:
#red color image 
img_R = img_array.copy()
#set green and blue intensity values of each pixel to 0.
img_R[:, :, (1, 2)] = 0

#green color image 
img_G = img_array.copy()
#set red and blue intensity values of each pixel to 0.
img_G[:, :, (0, 2)] = 0

#blue color image
img_B = img_array.copy()
#set red and green intensity values of each pixel to 0.
img_B[:, :, (0, 1)] = 0

concatenated_imgArray = np.concatenate((img_R, img_G, img_B), axis=1)

#save NumPy array as image file
pil_img = Image.fromarray(concatenated_imgArray)
pil_img.save('/content/sample_data/single_color_images.jpg')

Output:

<img src='https://drive.google.com/uc?id=1pSRnMgHoMmcwN6JJvlWAVcmHr-_tuYAn' width="600" height="160">

###Generate negative-positive inverted images (invert pixel value)

You can generate negative-positive inverted image by subtracting the pixel value from the maximum value.

As mentioned above, the data type of the array is uint8 and the maximum value of uint8 data type is 255. Therefore, we have to the pixel values from 255.

In [33]:
img_inverted = 255 - img_array

#save NumPy array as image file
pil_img = Image.fromarray(img_inverted)
pil_img.save('/content/sample_data/inverted_image.jpg')

Output:

<img src='https://drive.google.com/uc?id=1g1Jde_UGSfIBKCDCV9nY0whVVD-ezs0n' width="300" height="250">

###Trim the image using array slicing

You can use **array slicing** to trim the image to a rectangle. 

You need to specify an area with slice.

In [47]:
img_trimmed = img_array[400:660, 330:650]
print("Shape of the orginial image", img_array.shape)
print("Shape of the trimmed image",img_trimmed.shape)

#save NumPy array as image file
pil_img = Image.fromarray(img_trimmed)
pil_img.save('/content/sample_data/trimmed_image.jpg')

Shape of the orginial image (768, 1024, 3)
Shape of the trimmed image (260, 320, 3)


Output:

<img src='https://drive.google.com/uc?id=1LtuXPOJ60X4h3wPsmpwpGUBqKL9Up-Jh' width="300" height="250">